src/VisualStudio/Xaml/Impl/Implementation/LanguageServer/Handler/Definitions/GoToDefinitionHandler.cs

```diff
@@ -0,0 +1,152 @@
+
+            var definitions = await xamlGoToDefinitionService.GetDefinitionsAsync(document, position, cancellationToken).ConfigureAwait(false);
+            foreach (var definition in definitions)
+            {
+                locations.AddRange(await this.GetLocationsAsync(definition, context, cancellationToken).ConfigureAwait(false));
+            }
+
+            return locations.ToArrayAndFree();
+        }
+
+        private async Task<LSP.Location[]> GetLocationsAsync(XamlDefinition definition, RequestContext context, CancellationToken cancellationToken)
+        {
+            var locations = ArrayBuilder<LSP.Location>.GetInstance();
+
+            if (definition is XamlSourceDefinition sourceDefinition)
+            {
+                Contract.ThrowIfNull(sourceDefinition.FilePath);
+
+                var document = context.Solution?.GetDocuments(ProtocolConversions.GetUriFromFilePath(sourceDefinition.FilePath)).FirstOrDefault();
+                if (document != null)
+                {
+                    var sourceText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                    var span = sourceDefinition.GetTextSpan(sourceText);
+                    if (span != null)
+                    {
+                        var location = await ProtocolConversions.TextSpanToLocationAsync(
+                                                    document,
+                                                    span.Value,
+                                                    isStale: false,
+                                                    cancellationToken).ConfigureAwait(false);
+                        locations.AddIfNotNull(location);
+                    }
+                }
+                else
+                {
+                    // Cannot find the file in solution. This is probably a file lives outside of the solution like generic.xaml
+                    // which lives in the Windows SDK folder. Try getting the SourceText from the file path.
+                    using var fileStream = new FileStream(sourceDefinition.FilePath, FileMode.Open, FileAccess.Read);
+                    var sourceText = SourceText.From(fileStream);
+                    var span = sourceDefinition.GetTextSpan(sourceText);
+                    if (span != null)
+                    {
+                        locations.Add(new LSP.Location
+                        {
+                            Uri = new Uri(sourceDefinition.FilePath),
+                            Range = ProtocolConversions.TextSpanToRange(span.Value, sourceText),
+                        });
+                    }
+                }
+            }
+            else if (definition is XamlSymbolDefinition symbolDefinition)
+            {
+                Contract.ThrowIfNull(symbolDefinition.Symbol);
+                var symbol = symbolDefinition.Symbol;
+
+                if (symbol.Locations.First().IsInMetadata && _metadataAsSourceFileService != null)
```

Comment:
>symbol.Locations.First() [](http://example.com/codeflow?start=20&length=24)

@dibarbet IS the first location always the definition or is there a better way to get the location for a symbol definition?

Reply:
Yeah we usually use first.  that being said this is what we do in non-LSP and we probably need to update our handler to do something similar
https://sourceroslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/Workspace/VisualStudioSymbolNavigationService.cs,67